### PR TITLE
refactor(core): add command builder

### LINKED ISSUE #1626: refactor(core): PoC on pass_local_client decomposition
The current state is that all commands are decorated with a `pass_local_client` decorator which ensures that the client command is passed to all renku operation commands which uses the client object to ensure certain command lifecycles flows (locking, storage handling, partial datasets management, etc). Instead of having 1 "god" object which can do all those things, we should build each command through a builder pattern and attach controllers that are needed for this command. 

This decoupling should enable us the following:
1) Locking should be controlled on the command level (this is currently not possible with `pass_local_client` based control flow)
2) Committing should be controlled on the code level, instead of the current import level
  2.1) Users using renku as a library should be able to control this in case of the operation batching
3) Achieve more clarity and reasoning of the entire command lifecycle

As a first step towards this decoupling we could:
1. Rework current `LocalClient` into a `ProjectContext` object
  1.1. Flush out the public interface of the ProjectContext 
  1.2. Ideally we want to move away from mixins here since certain implementation should be easily replaceable (e.g. StorageApiMixin) - a good candidate for builder pattern as well
  1.3. Each existing mixin should be a controller which can be plugged and played into the context
  1.4. Interdependencies between mixing which we currently have should be destructed through the invocation parameters
2. Rework `pass_local_client` to be a builder pattern, such that:

```
@pass_local_client(clean=False, requires_migration=True, commit=True, commit_only=DATASET_METADATA_PATHS)
def dataset_remove(client, name, commit_message=None):
```

would become
```
dataset_remove = Command(
  clean=False, 
  requires_migration=True, 
  commit=True, 
  commit_only=DATASET_METADATA_PATHS
)
.set_action(dataset_remove)
.set_data({"name": name})
.set_lock({"project_id": "myproject"})
.build() <- this would build the needed context and inject it as a first parameter into the action implementation, this way we don't have to touch existing implementations.
```

after the command is built, the control flow should be exported to the client:
```
with ...:
   result = dataset_remove("my-dataset").execute(auto_commit=False)
   result.commit("chore: unused stuff removed")
```
Here `.execute` will create a lock file in `.renku/locks` director with the name of the command and invoke the set action and return the result object, which will embed the used context above and the reference will persist so that we can gain control over the `.commit`. The result object can also keep track of the `auto_commit` flag, so if the user sets it to auto_commit the operation will default to the current behavior but we gain control to batch operations.

This will enable us down the road to gain the batch operations on the API module. 

For example:
```
from renku.api import Project

with Project(".") as project:
  project.dataset_remove("my-dataset", auto_commit=False)
  project.dataset_remove("another-dataset", auto_commit=False)
  project.commit("removed all unnecessary datasets")
```

This ticket is about creating a PoC in which `pass_local_client` control flow should be decomposed into a manageable structure. As a PoC, change (1) can be done easily with the existing structure, for (2) we should introduce a new module and re-do the dataset create and remove with the new control flow. A good candidate for a new module would be: `renku.api` or `renku.core.incubation`.

### PR DESCRIPTION
**Pull request by Panaetius**
Adds an `renku.core.incubation` namespace with a command builder that injects a `LocalClient` into a wrapped command. Should easily support replacing `LocalClient` in the future (At least in this regard).

Open questions:
- Operation precedence: Some things like lock or creating a commit need a LocalClient, so a client needs to be created first. And we want to check if a migration is required before giving an exception for e.g. the repository not being clean. I tried to solve this with the different pre-/post-hooks having a priority and hooks being executed in descending priority, this way we can ensure the hooks are called in the right order, without having too strong coupling between hooks. There are other possible solutions to dealing with this, so we need to decide if we want to go with a hook based approach or something else.
- How far do we want to go right now? The hooks wrap the old pass_local_client snippets, but those could easily be replaced by proper pre/post hooks (then we don't need to do the contextmanager/stack.enter things we do right now).
- Should we switch `pass_local_client` over to this new implementation already?
- Do we need a more complex `CommandResult` interface?
- Do we already want to work on more complex locking? I added a dummy `DatasetLock` but without an implementation.
- how do we deal with cases where the builder needs an argument from the command? This is the case for Dataset locks, where we set the hook at command build time, but only know the dataset name at execution time. We could just say "The first argument of the command must be the dataset name" but that is a bit flimsy.

Other next steps we should/could do:
- allow pre-hooks to add/modify/remove args/kwargs of the command directly.
- add typing
- add an easier way for builder to specify prerequisites? Like "This builder requires this other builder to be used".

Closes #1626

### CODE REVIEW

File: tests/core/incubation/test_command.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        .raise_if_empty()
+        .lock_project()
+        .build()
+    )
+
+    create_dataset("ds1", title="", description="", creators=[], commit_message="my awesome dataset")
+    data_file = tmpdir / Path("somefile")
+    data_file.write_text("1,2,3", encoding="utf-8")
+
+    add_to_dataset.execute([str(data_file)], "ds1")
```
Comment (jsam) [inline tests/core/incubation/test_command.py:44]: This should return the result object? Should we catch it and assert its state?

**Review by jsam — COMMENTED** [review]
Look's amazing already! Thank you :rocket: 

I think we should add some `NOTE:` doc statements explaining the priorities (what are they used for and maybe small comment on what are the expected previous/next priorities in the life cycle). 

Next, I think with this new structure we should come up with new testing structure and define which things we should test for. For example, each command test could check for following:
1) Expected side-effect
2) Result object state
3) Lock management
4) Exceptions which can bubble up

File: renku/core/incubation/command.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+from renku.core.management.migrate import check_for_migration
+from renku.core.management.repository import default_path
+
+
+def check_finalized(f):
+    """Decorator to prevent modification of finalized builders."""
+
+    def wrapper(*args, **kwargs):
+        """Check finalized status."""
+        if args[0].finalized:
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/incubation/command.py:38]: only if we add a hook that's not a class method. But I can add a check for it and throw an appropriate exception.

**Review by mohammad-sdsc — APPROVED** [review]
Thanks Ralf for changes! Looks great!

**Review by mohammad-sdsc — APPROVED** [review]
LGTM!

**Review by jsam — APPROVED** [review]
Thanks! :rocket:

**Pull request merged**